### ISSUE TITLE
Add support for arm64e ABI (when possible)

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#4490]
See http://lists.llvm.org/pipermail/llvm-dev/2019-October/136091.html

```
arm64e is a specific ABI for pointer authentication in C, C++, 
Objective-C, and Swift.  It is used on iOS 12+ on systems that provide 
ARMv8.3, as well as several other Apple OSes.  This ABI is not yet 
considered “stable” for third-party use because some of the details 
are still evolving
```

This will likely be necessary in order to work with versions of Xcode (etc) that enable the pointer-authentication features for newer (A12+) chips.

Note that ARMv8.3a support is somewhat orthogonal to this; arm64e presumes/requires ARM8.3a to be available, but the reverse it not necessarily the case, AFAIK.

**Comment by steven-johnson** [comment]
See also https://github.com/halide/Halide/issues/4487

**Comment by rjmccall** [comment]
Yes, if you support emitting code for iOS (AOT, I guess, given iOS platform restrictions), you will need to support emitting code for pointer authentication.  If you don't need to generate any indirect calls, that should be very easy — just a small amount of driver setup.  Otherwise, it might be more complex.

**Comment by f0enix** [comment]
@steven-johnson I am trying to generate arm64e static library for my generators for ios, but it is not working. I tried using your proposed armv8.3a PR as a starting point but the final architecture is always arm64.

I added more details in my comment:
https://github.com/halide/Halide/pull/4489#issuecomment-846941511

I know this issue is old and many things may have changed in Halide meanwhile but i would like to know whether it is possible to generate arm64e static libraries from my generators.

**Comment by steven-johnson** [comment]
Adding @abadams @alexreinking -- we have a request internal to Google about supporting this (I'm rather surprised it hasn't come up before now, in hindsight). My guess is that if LLVM supports it directly (which I'd hope to be the case by now) then this should be fairly straightforward; if not.... well, not?

**Comment by abadams** [comment]
I think it hasn't come up because Apple backed off on requiring it. I don't know how complex it would be to add. It might just work, but it's also possible there could be LLVM IR passes we have to do over the runtime modules, because target-agnostic bitcode isn't enough.

**Comment by rjmccall** [comment]
The iOS JIT entitlement requires arm64e.  Users could also potentially want to run your code in an arm64e process on macOS for security reasons.  It's not urgent If you don't have users actually asking for these things, of course.

**Comment by steven-johnson** [comment]
As of July 2024, the support for arm64e in upstream LLVM doesn't appear to match that in recent XCode versions, so it's not really useful. This will likely need to wait for Apple to upstream their fixes from their private branch into the main branch.
